common/producer/BaseProducer.java

```diff
@@ -20,42 +20,53 @@
 
 import grakn.core.common.iterator.ResourceIterator;
 
+import javax.annotation.concurrent.ThreadSafe;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 import static grakn.core.common.concurrent.ExecutorService.forkJoinPool;
 
+@ThreadSafe
 public class BaseProducer<T> implements Producer<T> {
 
     private final ResourceIterator<T> iterator;
+    private final AtomicBoolean isDone;
     private CompletableFuture<Void> future;
 
     BaseProducer(ResourceIterator<T> iterator) {
         this.iterator = iterator;
+        this.isDone = new AtomicBoolean(false);
         this.future = CompletableFuture.completedFuture(null);
     }
 
     @Override
-    public synchronized void produce(Queue<T> queue, int count) {
-        future = future.thenRunAsync(() -> produceAsync(queue, count), forkJoinPool());
+    public synchronized void produce(Queue<T> queue, int request) {
+        if (isDone.get()) return;
+        future = future.thenRunAsync(() -> {
+            try {
+                int i = 0;
+                for (; i < request && iterator.hasNext() && !isDone.get(); i++) queue.put(iterator.next());
+                if (i < request && !isDone.get()) done(queue);
```

Comment:
I think we can do

```
for (; request > 0 && iterator.hasNext() && !isDone.get(); request--) queue.put(iterator.next());
if (request > 0 && !isDone.get()) done(queue);
```

Reply:
I've taken inspiration from this and do this instead:
```
int unfulfilled = request;
if (runningJobs.containsKey(iterator)) {
    for (; unfulfilled > 0 && iterator.hasNext() && !isDone.get(); unfulfilled--) {
        queue.put(iterator.next());
    }
}
if (!isDone.get()) transition(queue, iterator, unfulfilled);
```

NOTE: we should not modify input arguments if we can afford to. We should treat them as final, and create a new variable in which we can modify.